File: lib/sift/cli/queue/add.rb
# frozen_string_literal: true

require "json"

module Sift
  module CLI
    module Queue
      class Add < Base
        command_name "add"
        summary "Add a new item to the review queue"
        examples(
          "sq add --text 'Review this function'",
          "sq add --diff changes.patch --metadata '{\"workflow\":\"analyze\"}'",
          "echo 'content' | sq add --stdin text"
        )

        def define_flags(parser, options)
          options[:sources] ||= []
          options[:metadata] ||= {}

          parser.on("--diff PATH", "Add diff source (repeatable)") do |path|
            options[:sources] << { type: "diff", path: path }
          end

          parser.on("--file PATH", "Add file source (repeatable)") do |path|
            options[:sources] << { type: "file", path: path }
          end

          parser.on("--text STRING", "Add text source (repeatable)") do |text|
            options[:sources] << { type: "text", content: text }
          end

          parser.on("--directory PATH", "Add directory source (repeatable)") do |path|
            options[:sources] << { type: "directory", path: path }
          end

          parser.on("--stdin TYPE", Sift::Queue::VALID_SOURCE_TYPES,
            "Read source content from stdin (#{Sift::Queue::VALID_SOURCE_TYPES.join("|")})") do |type|
            options[:stdin_type] = type
          end

          parser.on("--title TITLE", "Title for the item") do |title|
            options[:title] = title
          end

          parser.on("--metadata JSON", "Attach metadata as JSON") do |json|
            options[:metadata] = parse_json(json, "metadata")
          end

          super
        end

        def execute
          if options[:stdin_type]
            content = $stdin.read
            options[:sources] << { type: options[:stdin_type], content: content }
          end

          if options[:sources].empty?
            logger.error("At least one source is required")
            logger.error("Use --diff, --file, --text, --directory, or --stdin")
            return 1
          end

          metadata = options[:metadata]

          item = queue.push(sources: options[:sources], title: options[:title], metadata: metadata)
          puts item.id
          logger.info("Added item #{item.id} with #{item.sources.length} source(s)")
          0
        end

        private

        def queue
          @queue ||= Sift::Queue.new(options[:queue_path])
        end

        def parse_json(str, field_name)
          JSON.parse(str)
        rescue JSON::ParserError => e
          raise OptionParser::InvalidArgument, "Invalid JSON for #{field_name}: #{e.message}"
        end
      end
    end
  end
end

File: sq-rust/src/cli/commands/add.rs
use crate::queue::{Queue, Source};
use crate::AddArgs;
use anyhow::Result;
use std::io::Read;
use std::path::PathBuf;

/// Execute the `sq add` command.
pub fn execute(args: &AddArgs, queue_path: PathBuf) -> Result<i32> {
    let queue = Queue::new(queue_path);

    let mut sources: Vec<Source> = Vec::new();

    for path in &args.diff {
        sources.push(Source {
            type_: "diff".to_string(),
            path: Some(path.clone()),
            content: None,
            session_id: None,
        });
    }

    for path in &args.file {
        sources.push(Source {
            type_: "file".to_string(),
            path: Some(path.clone()),
            content: None,
            session_id: None,
        });
    }

    for text in &args.text {
        sources.push(Source {
            type_: "text".to_string(),
            path: None,
            content: Some(text.clone()),
            session_id: None,
        });
    }

    for path in &args.directory {
        sources.push(Source {
            type_: "directory".to_string(),
            path: Some(path.clone()),
            content: None,
            session_id: None,
        });
    }

    if let Some(ref stdin_type) = args.stdin {
        let mut content = String::new();
        std::io::stdin().read_to_string(&mut content)?;
        sources.push(Source {
            type_: stdin_type.clone(),
            path: None,
            content: Some(content),
            session_id: None,
        });
    }

    if sources.is_empty() {
        eprintln!("Error: At least one source is required");
        eprintln!("Use --diff, --file, --text, --directory, or --stdin");
        return Ok(1);
    }

    let metadata = match &args.metadata {
        Some(json_str) => match serde_json::from_str(json_str) {
            Ok(v) => v,
            Err(e) => {
                eprintln!("Error: Invalid JSON for metadata: {}", e);
                return Ok(1);
            }
        },
        None => serde_json::Value::Object(serde_json::Map::new()),
    };

    let blocked_by: Vec<String> = match &args.blocked_by {
        Some(ids) => ids
            .split(',')
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty())
            .collect(),
        None => Vec::new(),
    };

    let item = queue.push(sources, args.title.clone(), metadata, None, blocked_by)?;
    println!("{}", item.id);
    eprintln!(
        "Added item {} with {} source(s)",
        item.id,
        item.sources.len()
    );
    Ok(0)
}
